File: src/com/example/javas/medium/M024_LoggerRateLimiter.java
package com.example.javas.medium;

import java.util.*;

/**
 * M024: Design Logger Rate Limiter
 * Difficulty: Medium (System Design)
 * Time Target: 25 minutes
 * 
 * Problem:
 * Design a logger system that receives a stream of messages along with their timestamps. 
 * Each unique message should only be printed at most every 10 seconds (i.e., a message 
 * printed at timestamp t will prevent other identical messages from being printed until 
 * timestamp t + 10).
 * 
 * All messages will come in chronological order. Several messages may arrive at the same timestamp.
 * 
 * Requirements:
 * Implement the Logger class:
 * - Logger() Initializes the logger object
 * - boolean shouldPrintMessage(int timestamp, String message)
 *   Returns true if the message should be printed in the given timestamp, otherwise false
 * 
 * Constraints:
 * - 0 <= timestamp <= 10^9
 * - Every timestamp will be in non-decreasing order (chronological order)
 * - 1 <= message.length <= 30
 * - At most 10^4 calls will be made to shouldPrintMessage
 * 
 * Examples:
 * Logger logger = new Logger();
 * 
 * logger.shouldPrintMessage(1, "foo");  // return true, print "foo" at timestamp 1
 * logger.shouldPrintMessage(2, "bar");  // return true, print "bar" at timestamp 2
 * logger.shouldPrintMessage(3, "foo");  // return false, "foo" was printed at t=1, only 2 seconds passed
 * logger.shouldPrintMessage(8, "bar");  // return false, "bar" was printed at t=2, only 6 seconds passed
 * logger.shouldPrintMessage(10, "foo"); // return false, "foo" was printed at t=1, only 9 seconds passed
 * logger.shouldPrintMessage(11, "foo"); // return true, "foo" last printed at t=1, 10 seconds passed
 * 
 * Approach:
 * Use a HashMap to track the last print timestamp for each message:
 * 1. Key: message string
 * 2. Value: last timestamp when this message was printed
 * 3. For each shouldPrintMessage call:
 *    - Check if message exists in map and if less than 10 seconds passed
 *    - If can print: update map with new timestamp and return true
 *    - If cannot print: return false
 * 
 * Follow-up (optional):
 * - What if messages can arrive out of order?
 * - How would you clean up old entries to prevent unlimited memory growth?
 * 
 * Data Structures:
 * - HashMap<String, Integer> to map message -> last print timestamp
 * 
 * Time Complexity: O(1) for shouldPrintMessage
 * Space Complexity: O(M) where M is number of unique messages
 */
public class M024_LoggerRateLimiter {
    
    static class Logger {
        
        // TODO: Declare a HashMap to store message and its last print timestamp
        
        public Logger() {
            // TODO: Initialize the HashMap
        }
        
        public boolean shouldPrintMessage(int timestamp, String message) {
            // TODO: Implement rate limiting logic
            // 1. Check if message exists in the map
            // 2. If message not in map OR (current timestamp - last timestamp) >= 10:
            //    - Update map with current timestamp
            //    - Return true (should print)
            // 3. Otherwise:
            //    - Return false (still in cooldown period)
            
            return false;
        }
    }
    
    // Test cases
    public static void main(String[] args) {
        // Test case 1: Basic functionality
        System.out.println("Test 1: Basic rate limiting");
        Logger logger1 = new Logger();
        System.out.println(logger1.shouldPrintMessage(1, "foo"));  // true
        System.out.println(logger1.shouldPrintMessage(2, "bar"));  // true
        System.out.println(logger1.shouldPrintMessage(3, "foo"));  // false (only 2 sec)
        System.out.println(logger1.shouldPrintMessage(8, "bar"));  // false (only 6 sec)
        System.out.println(logger1.shouldPrintMessage(10, "foo")); // false (only 9 sec)
        System.out.println(logger1.shouldPrintMessage(11, "foo")); // true (10 sec passed)
        
        // Test case 2: Multiple different messages
        System.out.println("\nTest 2: Multiple messages");
        Logger logger2 = new Logger();
        System.out.println(logger2.shouldPrintMessage(0, "a"));    // true
        System.out.println(logger2.shouldPrintMessage(0, "b"));    // true
        System.out.println(logger2.shouldPrintMessage(0, "c"));    // true
        System.out.println(logger2.shouldPrintMessage(5, "a"));    // false
        System.out.println(logger2.shouldPrintMessage(10, "a"));   // true
        System.out.println(logger2.shouldPrintMessage(10, "b"));   // true
        
        // Test case 3: Same message at exact 10 second boundary
        System.out.println("\nTest 3: Exact 10 second boundary");
        Logger logger3 = new Logger();
        System.out.println(logger3.shouldPrintMessage(5, "test"));  // true
        System.out.println(logger3.shouldPrintMessage(14, "test")); // false (only 9 sec)
        System.out.println(logger3.shouldPrintMessage(15, "test")); // true (exactly 10 sec)
        
        // Test case 4: Messages at same timestamp
        System.out.println("\nTest 4: Same timestamp");
        Logger logger4 = new Logger();
        System.out.println(logger4.shouldPrintMessage(1, "x"));  // true
        System.out.println(logger4.shouldPrintMessage(1, "x"));  // false (same time)
        System.out.println(logger4.shouldPrintMessage(1, "y"));  // true (different msg)
    }
}

File: pythonX/medium/m024_logger_rate_limiter.py
"""
M024: Design Logger Rate Limiter

Summary:
Design logger that prints messages at most every 10 seconds.

Requirements:
Implement Logger class:
- should_print_message(timestamp, message): Return True if should print

Approach:
Use dict to track last print timestamp for each message.

Data Structures:
- dict[str, int]: message -> last print timestamp

Big O Analysis:
Time Complexity: O(1)
Space Complexity: O(M) where M is unique messages
"""


class Logger:
    
    def __init__(self):
        """
        TODO: Initialize dict to store message -> last timestamp
        """
        pass
    
    def should_print_message(self, timestamp: int, message: str) -> bool:
        """
        TODO: Check if message should be printed
        1. Check if message in dict
        2. If not in dict OR (timestamp - last_timestamp) >= 10:
           - Update dict with timestamp
           - Return True
        3. Otherwise return False
        """
        pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Basic rate limiting")
    logger1 = Logger()
    print(logger1.should_print_message(1, "foo"))   # True
    print(logger1.should_print_message(2, "bar"))   # True
    print(logger1.should_print_message(3, "foo"))   # False (only 2 sec)
    print(logger1.should_print_message(8, "bar"))   # False (only 6 sec)
    print(logger1.should_print_message(10, "foo"))  # False (only 9 sec)
    print(logger1.should_print_message(11, "foo"))  # True (10 sec passed)
    
    print("\nTest 2: Multiple messages")
    logger2 = Logger()
    print(logger2.should_print_message(0, "a"))     # True
    print(logger2.should_print_message(0, "b"))     # True
    print(logger2.should_print_message(0, "c"))     # True
    print(logger2.should_print_message(5, "a"))     # False
    print(logger2.should_print_message(10, "a"))    # True
    print(logger2.should_print_message(10, "b"))    # True
    
    print("\nTest 3: Exact 10 second boundary")
    logger3 = Logger()
    print(logger3.should_print_message(5, "test"))  # True
    print(logger3.should_print_message(14, "test")) # False (9 sec)
    print(logger3.should_print_message(15, "test")) # True (exactly 10)
    
    print("\nTest 4: Same timestamp")
    logger4 = Logger()
    print(logger4.should_print_message(1, "x"))     # True
    print(logger4.should_print_message(1, "x"))     # False (same time)
    print(logger4.should_print_message(1, "y"))     # True (different msg)
